### ISSUE TITLE
Support nested associative arrays

### ISSUE THREAD

**Issue by dantleech** [doctrine/phpcr-odm#417]
It should be possible to have an array type which automatically creates a structure, e.g

``` php
class Site {
    /**
     * @PHPCR\AssocArray()
     */
     protected $preferences;
}
```

Then automatically map that to a PHPCR structure:

```
/site
    preferences/
         - jcr:primaryType  = phpcr:associative_array
         - key1 = value 
         key2/
             - key3: = value
             - key4: = value
             key5/
```

**Comment by dantleech** [comment]
I just wrote a an event listener class for SuluCMF which does this. It uses a dot-delimited property name to store multi-dimensional arrays.

| Name | Type | Vaue |
| --- | --- | --- |
| foo.bar.bar  | STRING  | Value 1 |
| foo.boo.baz | LONG | 1234 |
| foo.car | STRING | Value 2 |

Would give

````
array(
    'foo' => array(
        'bar' => => array(
            'bar' => 'Value 1',
            'baz' => 1234,
        ),
        'car' => 'Value 2',
    ),
````
    
Will try and port it here.

**Comment by dbu** [comment]
so you use a child node for the array, but only one instead of a nested node tree? i guess that works, but why not go with nested nodes while you are at it? this scheme would break if i do 

```
array(
    'foo.bar' => 'x',
);
```

**Comment by dantleech** [comment]
No everything is in the same node in this strategy.

Regarding the breaking of the scheme, if we rewrote the entire array every time it shouldn't be a problem I think.

**Comment by dbu** [comment]
if everything is in the same node, does that mean there can only be one array? and how do you know what belongs into the array or not?

with rewriting, you mean encode `.` in hashmap keys with something else?

**Comment by dantleech** [comment]
No, say you have a property `foo` which is mapped as an `array` in PHPCR-ODM.

If you set its value to `array('foo' => array('bar' => 'baz'), 'car' => array('boat', 'plane'))` then the following properties would be created in the node:

````
foo.foo.bar STRING "baz"
foo.car.0 STRING "boat"
foo.car.1 STRING "plane"
````

**Comment by dantleech** [comment]
> with rewriting, you mean encode . in hashmap keys with something else?

Ah, yeah I see what you mean. We would need to encode/escape the delimiter.

**Comment by dbu** [comment]
ah ok. i think your first example was missing the property name, that got me confused ;-)

and then to rebuild the array, you would scan the node for all properties starting with "foo.". but be aware that properties do not have guaranteed order, only nodes can have that. this means we can not guarantee to rebuild the exact same php array. but this is no argument against your architecture, as the same problem holds for nested nodes for child arrays.

the order would only be guaranteed if we would use a single multivalue field with a list of keys and a second one with a list of values. but then it would become unsearchable again. we might use a separate multivalue field just for the order.

**Comment by dantleech** [comment]
Hmm yeah. But is order really important with associative arrays?

In the case of non-associative arrays they would have indexes so would be orderable. I think unordered associative arrays is an acceptable compromise (although if pushed we could store the order as you say).

**Comment by dbu** [comment]
well, i could do a foreach on the assoc array... i think while we are at it we should make it clean, or we risk unexpected things. the separate multivalue field for the order should be ok. you could then restore the array by doing foreach on that multivalue field and get the values out of the property values array of the node.

**Comment by dantleech** [comment]
maybe we could make it an option?

````php
/**
 * @PHPCR\Array(ordered=true)
 */
protected $options;
````

Storing the order would increase the node size, and this strategy is already pretty inefficient in terms of size -- although perhaps this is a moot point anyway as we need to store the keys in order to have the `null` values.

Also .. this could replace the existing array implementation(s) what do you think?

**Comment by dantleech** [comment]
(regarding the `null` values we could also represent them as a string e.g. `__null__` and make sure that that value is escaped when the user uses it) /cc @uwej711

**Comment by dbu** [comment]
i would not make keeping the order optional - keep complexity manageable. if anything, we could support either expanding the way you do, or serializing an array (if you are concerned about efficiency).

and yes, this should probably replace what uwe did, no need to have two slightly different things for the same. this however is a BC break for existing data...

can you search the same way as with the uwe-solution? right now if i had a field $f, it was f = 'foo', now it would be f.* = 'foo' right?

**Comment by dantleech** [comment]
I guess for searching we should make it transparent (and therefore maintain BC). I guess we need to do some research about the best way to implement the search syntax/API. But I think the basic case `f = 'foo'` should be the same.

**Comment by dbu** [comment]
in the query builder, yes (though, you could also want to search for a specific key?)

i was thinking about SQL2 queries too, there must be a representation that works.

**Comment by dantleech** [comment]
well, I guess SQL2 is to PHPCR-ODM what SQL is to Doctrine ORM.. i.e. it breaks the abstraction so you shouldn't really use it to query for documents.

**Comment by dantleech** [comment]
Had a look at this. Basically there is a pain point with the translation strategies because the they would need to set a property value for each element in the array instead of a single value. So we would probably have to push more responsiblity into those (although I guess it could be handled in an abstract class and we would have a "serialization" helper).

**Comment by dbu** [comment]
for child strategy, we need not do anything at all, as each translation has its own node. for attribute, we should probably just translate each property, if you go with the approach of splitting each array entry into its own property.
